cmd/kube-apiserver/app/server.go

```diff
@@ -583,7 +584,19 @@ func BuildStorageFactory(s *options.ServerRunOptions) (*serverstorage.DefaultSto
 		storageFactory.SetEtcdLocation(groupResource, servers)
 	}
 
-	if s.Etcd.EncryptionProviderConfigFilepath != "" {
+	// Set up encryption at rest via transformer overrides if a configuration file was provided.
+	if len(s.Etcd.EncryptionProviderConfigFilepath) != 0 {
+		cloud, err := cloudprovider.InitCloudProvider(s.CloudProvider.CloudProvider, s.CloudProvider.CloudConfigFile)
+		if err != nil {
+			return nil, err
+		}
+		// Returns an envelope Service implemented by a KMS supported by the cloud hosting k8s.
+		cloudKMSGetter := func(name string) (envelope.Service, error) {
+			return cloud.KeyManagementService(name)
+		}
+		// Register cloud in plugin registry in case a cloudprovided KMS provider is required.
+		encryptionconfig.KMSPluginRegistry.RegisterCloudProvidedKMSPlugin(cloudKMSGetter)
```

Comment:
Because we don't know the name of the cloudprovided plugins in the kube-apiserver. For instance, we don't have access to "google-cloudkms" string here.
We have 2 different ways to configure KMS based encryption providers, `cloudprovidedkms` and `kms`. Mixing them up may not be clean.
Can you suggest something different to avoid mutating the global singleton?